UnwrapBlog.md

```diff
@@ -0,0 +1,92 @@
+
+Up till now, we'd supplied some simple helpers to make this more managable.  Our helpers would effectively 'unwrap' one layer of this sort of structure allowing you take the external value and deal with an internal value of a known shape.  So it effectively worked like this:
+
+```ts
+function unwrap<T>(input: Input<T>): T { /* ... */ }
+
+// so, you could then write:
+const ipAddresses: Input<string>[] = unwrap(vpc.ipAddresses);
+```
+
+This helped things somewhat.  But it still meant that as you processed that array you had to then 
+unwrap each and every element.  As our data model has grown, and we've tackled more and more
+complex domains, we started seeing code start to form the programming [Pyramid of Doom](https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)), where you just kept on going deeper and deeper, unwrapping all the way.  We started seeing this happen so often that we wanted to see if we there was anything we could do to make life much nicer for people using our APIs and using these helpers.
+
+Our core goal was to change our 'unwrap' function from just peeling away one layer of this onion, to peeling away *all* layers.  So, for example, if you started with an `Input<Input<string>[]>` you would just get back a plain old `string[]` that you could use completely naturally in all the ways you'd expect.  We were initially uncertain if we could make this work, but thanks to the power of JavaScript and the expressivity of TypeScript, we think we found a great solution here.
+
+We knew that providing the implementation such a function itself was not going to the true hard part.  After all, we already had the code to peel back one layer.  So peeling back all the layers was more a matter of just having the function call itself recursively when encountering things like arrays (or other objects that themselves contained `Input`s).  The hard part came in when were trying figure out how would we even express this in TypeScript?.  After all, what's the type signature when you effectively to say:
+
+```ts
+function unwrap(input: Input<...MaybeHasInputsOrPromisesOrOutputsArbitrarilyDeeper...>): SomeThingWithAllTheInputsPromisesAndOutputsErased;
+```
+
+This was a big headscratcher for a while.  But after exploring some of the powerful new features that TypeScript has been bringing about in recent releases, we finally found a great way to manage it.  In the end, it ends up looking something like this:
```

Comment:
Any way to link "powerful new features" to release notes or examples? E.g., for this sort of thing, I'd usually link powerful, new, and features to different examples of our favorite features here.

Reply:
Actually, I see that you link below, so this is probably okay as-is.